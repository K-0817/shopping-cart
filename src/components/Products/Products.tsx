import { IProduct } from 'models';
import Product from './Product';

import * as S from './style';

interface IProps {
  products: IProduct[];
}

const Products = ({ products }: IProps) => {
  return (
    <S.Container>
      {products?.map((p) => (
        <Product product={p} key={p.sku} />
      ))}
    </S.Container>
  );
};

export default Products;
