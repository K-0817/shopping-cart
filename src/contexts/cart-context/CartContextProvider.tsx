import { createContext, useContext, FC, useState } from 'react';
import { ICartProduct, ICartTotal } from 'models';

export interface ICartContext {
  isOpen: boolean;
  setIsOpen(state: boolean): void;
  products: ICartProduct[];
  setProducts(products: ICartProduct[]): void;
  total: ICartTotal;
  setTotal(products: any): void;
}

const CartContext = createContext<ICartContext | undefined>(undefined);
const useCartContext = (): ICartContext => {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error('useCartContext must be used within a CartProvider');
  }

  return context;
};

const totalInitialValues = {
  productQuantity: 0,
  installments: 0,
  totalPrice: 0,
  currencyId: 'USD',
  currencyFormat: '$',
};

const CartProvider: FC = (props) => {
  const [isOpen, setIsOpen] = useState(false);
  const [products, setProducts] = useState<ICartProduct[]>([]);
  const [total, setTotal] = useState<ICartTotal>(totalInitialValues);

  const CartContextValue: ICartContext = {
    isOpen,
    setIsOpen,
    products,
    setProducts,
    total,
    setTotal,
  };

  return <CartContext.Provider value={CartContextValue} {...props} />;
};

export { CartProvider, useCartContext };
