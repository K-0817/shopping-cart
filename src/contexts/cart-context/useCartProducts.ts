import { useCartContext } from './CartContextProvider';
import useCartTotal from './useCartTotal';
import { ICartProduct } from 'models';

const useCartProducts = () => {
  const { products, setProducts } = useCartContext();
  const { updateCartTotal } = useCartTotal();

  const updateQuantitySafely = (
    currentProduct: ICartProduct,
    targetProduct: ICartProduct,
    quantity: number
  ): ICartProduct => {
    if (currentProduct.id === targetProduct.id) {
      return Object.assign({
        ...currentProduct,
        quantity: currentProduct.quantity + quantity,
      });
    } else {
      return currentProduct;
    }
  };

  const addProduct = (newProduct: ICartProduct) => {
    let updatedProducts;
    const isProductAlreadyInCart = products.some(
      (product: ICartProduct) => newProduct.id === product.id
    );

    if (isProductAlreadyInCart) {
      updatedProducts = products.map((product: ICartProduct) => {
        return updateQuantitySafely(product, newProduct, newProduct.quantity);
      });
    } else {
      updatedProducts = [...products, newProduct];
    }

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  const removeProduct = (productToRemove: ICartProduct) => {
    const updatedProducts = products.filter(
      (product: ICartProduct) => product.id !== productToRemove.id
    );

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  const increaseProductQuantity = (productToIncrease: ICartProduct) => {
    const updatedProducts = products.map((product: ICartProduct) => {
      return updateQuantitySafely(product, productToIncrease, +1);
    });

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  const decreaseProductQuantity = (productToDecrease: ICartProduct) => {
    const updatedProducts = products.map((product: ICartProduct) => {
      return updateQuantitySafely(product, productToDecrease, -1);
    });

    setProducts(updatedProducts);
    updateCartTotal(updatedProducts);
  };

  return {
    products,
    addProduct,
    removeProduct,
    increaseProductQuantity,
    decreaseProductQuantity,
  };
};

export default useCartProducts;
