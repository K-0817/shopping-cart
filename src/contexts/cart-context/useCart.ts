import { useCartContext } from './CartContextProvider';
import useCartProducts from './useCartProducts';
import useCartTotal from './useCartTotal';

const useCart = () => {
  const { isOpen, setIsOpen } = useCartContext();
  const {
    products,
    addProduct,
    removeProduct,
    increaseProductQuantity,
    decreaseProductQuantity,
  } = useCartProducts();
  const { total, updateCartTotal } = useCartTotal();

  const openCart = () => setIsOpen(true);
  const closeCart = () => setIsOpen(false);

  return {
    isOpen,
    openCart,
    closeCart,
    products,
    addProduct,
    removeProduct,
    increaseProductQuantity,
    decreaseProductQuantity,
    total,
    updateCartTotal,
  };
};

export default useCart;
