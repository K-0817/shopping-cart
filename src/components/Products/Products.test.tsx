import { renderWithThemeProvider } from 'utils/test/test-utils';
import { mockProducts } from 'utils/test/mocks';

import { CartProvider } from 'contexts/cart-context';
import Products from '.';

describe('[components] - Products', () => {
  const setup = (props = {}) => {
    return renderWithThemeProvider(
      <CartProvider>
        <Products products={mockProducts} {...props} />
      </CartProvider>
    );
  };

  test('should render correctly', () => {
    const view = setup();
    expect(view).toMatchSnapshot();
  });
});
