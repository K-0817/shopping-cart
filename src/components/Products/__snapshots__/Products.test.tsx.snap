// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] - Products should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Products__Container-sc-uhudcz-0 kTCRJn"
      >
        <div
          class="Product__Container-sc-124al1g-2 kgHzaj"
          tabindex="1"
        >
          <div
            class="Product__Stopper-sc-124al1g-3 iirFfh"
          >
            Free shipping
          </div>
          <div
            alt="Cropped Stay Groovy off white"
            class="Product__Image-sc-124al1g-1"
          />
          <p
            class="Product__Title-sc-124al1g-4 iKNVRs"
          >
            Cropped Stay Groovy off white
          </p>
          <div
            class="Product__Price-sc-124al1g-5 iZlLwH"
          >
            <p
              class="Product__Val-sc-124al1g-6 eKyHNu"
            >
              <small>
                $
              </small>
              <b>
                10
              </b>
              <span>
                .90
              </span>
            </p>
            <p
              class="Product__Installment-sc-124al1g-7 hBZpTG"
            >
              <span>
                or 
                9
                 x
              </span>
              <b>
                $
                1.21
              </b>
            </p>
          </div>
          <button
            class="Product__BuyButton-sc-124al1g-0 jCZDKU"
            tabindex="-1"
          >
            Add to cart
          </button>
        </div>
        <div
          class="Product__Container-sc-124al1g-2 rtefv"
          tabindex="1"
        >
          <div
            class="Product__Stopper-sc-124al1g-3 iirFfh"
          >
            Free shipping
          </div>
          <div
            alt="Basic Cactus White T-shirt"
            class="Product__Image-sc-124al1g-1"
          />
          <p
            class="Product__Title-sc-124al1g-4 iKNVRs"
          >
            Basic Cactus White T-shirt
          </p>
          <div
            class="Product__Price-sc-124al1g-5 iZlLwH"
          >
            <p
              class="Product__Val-sc-124al1g-6 eKyHNu"
            >
              <small>
                $
              </small>
              <b>
                13
              </b>
              <span>
                .25
              </span>
            </p>
            <p
              class="Product__Installment-sc-124al1g-7 hBZpTG"
            >
              <span>
                or 
                3
                 x
              </span>
              <b>
                $
                4.42
              </b>
            </p>
          </div>
          <button
            class="Product__BuyButton-sc-124al1g-0 jCZDKU"
            tabindex="-1"
          >
            Add to cart
          </button>
        </div>
        <div
          class="Product__Container-sc-124al1g-2 rowhT"
          tabindex="1"
        >
          <div
            alt="Skater Black Sweatshirt"
            class="Product__Image-sc-124al1g-1"
          />
          <p
            class="Product__Title-sc-124al1g-4 iKNVRs"
          >
            Skater Black Sweatshirt
          </p>
          <div
            class="Product__Price-sc-124al1g-5 iZlLwH"
          >
            <p
              class="Product__Val-sc-124al1g-6 eKyHNu"
            >
              <small>
                $
              </small>
              <b>
                25
              </b>
              <span>
                .90
              </span>
            </p>
            <p
              class="Product__Installment-sc-124al1g-7 hBZpTG"
            >
              <span>
                or 
                12
                 x
              </span>
              <b>
                $
                2.16
              </b>
            </p>
          </div>
          <button
            class="Product__BuyButton-sc-124al1g-0 jCZDKU"
            tabindex="-1"
          >
            Add to cart
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Products__Container-sc-uhudcz-0 kTCRJn"
    >
      <div
        class="Product__Container-sc-124al1g-2 kgHzaj"
        tabindex="1"
      >
        <div
          class="Product__Stopper-sc-124al1g-3 iirFfh"
        >
          Free shipping
        </div>
        <div
          alt="Cropped Stay Groovy off white"
          class="Product__Image-sc-124al1g-1"
        />
        <p
          class="Product__Title-sc-124al1g-4 iKNVRs"
        >
          Cropped Stay Groovy off white
        </p>
        <div
          class="Product__Price-sc-124al1g-5 iZlLwH"
        >
          <p
            class="Product__Val-sc-124al1g-6 eKyHNu"
          >
            <small>
              $
            </small>
            <b>
              10
            </b>
            <span>
              .90
            </span>
          </p>
          <p
            class="Product__Installment-sc-124al1g-7 hBZpTG"
          >
            <span>
              or 
              9
               x
            </span>
            <b>
              $
              1.21
            </b>
          </p>
        </div>
        <button
          class="Product__BuyButton-sc-124al1g-0 jCZDKU"
          tabindex="-1"
        >
          Add to cart
        </button>
      </div>
      <div
        class="Product__Container-sc-124al1g-2 rtefv"
        tabindex="1"
      >
        <div
          class="Product__Stopper-sc-124al1g-3 iirFfh"
        >
          Free shipping
        </div>
        <div
          alt="Basic Cactus White T-shirt"
          class="Product__Image-sc-124al1g-1"
        />
        <p
          class="Product__Title-sc-124al1g-4 iKNVRs"
        >
          Basic Cactus White T-shirt
        </p>
        <div
          class="Product__Price-sc-124al1g-5 iZlLwH"
        >
          <p
            class="Product__Val-sc-124al1g-6 eKyHNu"
          >
            <small>
              $
            </small>
            <b>
              13
            </b>
            <span>
              .25
            </span>
          </p>
          <p
            class="Product__Installment-sc-124al1g-7 hBZpTG"
          >
            <span>
              or 
              3
               x
            </span>
            <b>
              $
              4.42
            </b>
          </p>
        </div>
        <button
          class="Product__BuyButton-sc-124al1g-0 jCZDKU"
          tabindex="-1"
        >
          Add to cart
        </button>
      </div>
      <div
        class="Product__Container-sc-124al1g-2 rowhT"
        tabindex="1"
      >
        <div
          alt="Skater Black Sweatshirt"
          class="Product__Image-sc-124al1g-1"
        />
        <p
          class="Product__Title-sc-124al1g-4 iKNVRs"
        >
          Skater Black Sweatshirt
        </p>
        <div
          class="Product__Price-sc-124al1g-5 iZlLwH"
        >
          <p
            class="Product__Val-sc-124al1g-6 eKyHNu"
          >
            <small>
              $
            </small>
            <b>
              25
            </b>
            <span>
              .90
            </span>
          </p>
          <p
            class="Product__Installment-sc-124al1g-7 hBZpTG"
          >
            <span>
              or 
              12
               x
            </span>
            <b>
              $
              2.16
            </b>
          </p>
        </div>
        <button
          class="Product__BuyButton-sc-124al1g-0 jCZDKU"
          tabindex="-1"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
